Scale-dependent mu-Sigma modified gravity support

mishakb opened the pull request:
This PR implements a scale-dependent modified gravity using mu(a,k)-Sigma(a,k) parameterization. The structure of this implementation is substantially different and expanded compared to the previous mu-Sigma implementations (PR #442) and (PR #737).  The new MG parameter functions are now implemented so they can be incorporated to the transfer functions where the MG dependence on k has been made possible by the addition of new routines. The new capabilities are illustrated in a preliminary Jupyter notebook including redshift and scale dependences.  The new capabilities cover the matter power spectrum, the 2D angular spectra and correlations, and CMB Lensing. The models can take a single wavenumber k or an array of k and return results according to the user input. If no scale input is provided, the redshift-only results and examples are all reproduced correctly.  I will add a full description of the new implementation features to the documentation and the CCL-note. 
Changes and additions of code and routines can be found in the following 8 files: 

include/ccl_background.h
pyccl/background.py
pyccl/ccl_background.i
pyccl/core.py
pyccl/pyutils.py
pyccl/tracers.py
src/ccl_background.c
src/ccl_core.c

I have also temporarily uploaded the two following Jupyter notebooks to the examples folder to have a look at the examples but will move them in time to the CCLX repo. 

MG_redshift_and_scale_dependences_examples.ipynb
MG_redshift_only_dependence_examples.ipynb

Work that remain to be done:
- Cleaning some commented code (and flake8 errors)
- Benchmarking
- Unit tests
- Documentation

While I will be working on the above, this is a good time to have some feedback.

c-d-leonard commented:
@mishakb still looking for feedback on this pre-formal review? If so I can take a look now.

mishakb commented:
> @mishakb still looking for feedback on this pre-formal review? If so I can take a look now.

@c-d-leonard,  that would be great! Thanks a lot!

c-d-leonard commented on pyccl/background.py:
```diff
@@ -271,35 +272,65 @@ def rho_x(cosmo, a, species, is_comoving=False):
-            modified gravity at scale factor a. \
-            Sig_MG is assumed to be proportional to Omega_Lambda(z), \
+        float or array_like: Modification to Poisson equation \
+            under modified gravity at a scale factor. \
```
Small note, I believe the original wording of this is intended to mean "at scale factor given by a", which I think gives more information than the new wording?

mishakb replied:
I see now the meaning -- fixed!

c-d-leonard commented on pyccl/background.py:
```diff
@@ -271,35 +272,65 @@ def rho_x(cosmo, a, species, is_comoving=False):
 
     Args:
-        cosmo (:class:`~pyccl.core.Cosmology`): a Cosmology object.
+        cosmo (:class:`~pyccl.core.Cosmology`): Cosmological parameters.
```
I think this is clearer as "a Cosmology object", to be explicit about the type of variable cosmo should be?

mishakb replied:
agreed. Fixed.

c-d-leonard commented on pyccl/ccl_background.i.old:
```diff
@@ -0,0 +1,150 @@
+%module ccl_background
```
We shouldn't commit this file, github will keep track of the old version for us.

mishakb replied:
slipped through. Removed!

mishakb replied:
now removed.

c-d-leonard commented on pyccl/tracers.py:
```diff
@@ -1,9 +1,68 @@
 import numpy as np
 
+def get_MG_transfer_function(cosmo, z, k_MG=None):
+    """ This function allows to obtain the function Sigma(z,k) (1 or 2 D vector)
```
Just to avoid confusion, I would say "1 or 2D array" because I think "vector" is commonly taken to be 1D only.

mishakb replied:
Changed!

c-d-leonard commented on pyccl/tracers.py:
```diff
@@ -1,9 +1,68 @@
+def get_MG_transfer_function(cosmo, z, k_MG=None):
+    """ This function allows to obtain the function Sigma(z,k) (1 or 2 D vector)
+        for an array redshift (i.e. scale factor) coming from the redshift 
+        distributiorn (defined by the user) and a single value or an array of k 
```
typo

mishakb replied:
fixed!

c-d-leonard commented on pyccl/tracers.py:
```diff
@@ -1,9 +1,68 @@
+    """
+	
+    if isinstance(z, float):
+        a_CMB=1/(1+z)
```
I would prefer to keep this naming more general (e.g. a_single), although I understand the CMB is the most probable use here.

mishakb replied:
That will work. Done!

c-d-leonard commented on pyccl/tracers.py:
```diff
@@ -262,6 +321,95 @@ def get_bessel_derivative(self):
+        """
+        # case with no astro biases
+        if ((bias_transfer_a is None) and (bias_transfer_k is None)):    
+            if (isinstance(k_MG, (list, np.ndarray))): 		
```
I don't really understand why you are splitting this into three if statements - looks like each case does the same thing? Am I missing something?

c-d-leonard replied:
Oh, I see the first one has a transfer_ka while the latter two have transfer_a - but the second two still seem to be the same?

mishakb replied:
Good point. In principle, that is right and I could condense the two last into one. The reason I did not is that for future development in case some further needs to treat these cases differently. I added a comment to explain this but I can do it if you feel it is best.

c-d-leonard replied:
Building in flexibility makes sense to me. Happy for you to leave it.

c-d-leonard commented on pyccl/tracers.py:
```diff
@@ -262,6 +321,95 @@ def get_bessel_derivative(self):
+                self.add_tracer(cosmo, kernel,
+                    transfer_ka=mg_transfer_new,
+                    der_bessel=der_bessel, der_angles=der_angles)	
+            elif (isinstance(k_MG, float) or isinstance(k_MG, int)):
```
Again, these second two cases seem to have the same outcome?

mishakb replied:
see response above.

c-d-leonard commented on pyccl/tracers.py:
```diff
@@ -262,6 +321,95 @@ def get_bessel_derivative(self):
+                    transfer_ka=mg_transfer_new,
+                     der_bessel=der_bessel, der_angles=der_angles)
+            elif (isinstance(k_MG, float) or isinstance(k_MG, int)):
+                # Scale-dependent case with  a single specified k value
```
second two cases look the same again here

mishakb replied:
see response above.

c-d-leonard commented on pyccl/tracers.py:
```diff
@@ -262,6 +321,95 @@ def get_bessel_derivative(self):
+                self.add_tracer(cosmo, kernel,
+                    transfer_ka=mg_transfer_new, 
+                    der_bessel=der_bessel, der_angles=der_angles)
+            elif (isinstance(k_MG, float) or isinstance(k_MG, int)):
```
same thing here

mishakb replied:
see response above.

c-d-leonard commented on readthedocs/api/pyccl.halos.rst:
```diff
@@ -7,12 +7,18 @@ Submodules
+   pyccl.halos.hbias
+   pyccl.halos.hmfunc
+   pyccl.halos.massdef
+=======
```
You have a merge conflict to fix here.

mishakb replied:
now removed

c-d-leonard commented on readthedocs/api/pyccl.rst:
```diff
@@ -7,7 +7,10 @@ Subpackages
+=======
     pyccl.nl_pt
+>>>>>>> master
 
```
merge conflict here

mishakb replied:
now removed

c-d-leonard commented on readthedocs/index.rst:
```diff
@@ -6,10 +6,19 @@ The Core Cosmology Library (CCL) is a standardized library of routines to calcul
 
 The core functions of this package include:
 
+<<<<<<< HEAD
```
merge conflict here.

mishakb replied:
now removed

c-d-leonard commented on readthedocs/source/installation.rst:
```diff
@@ -12,6 +12,8 @@ as well. See the instructions for :ref:`boltzmann-codes` and
 
 CCL works on Linux or Mac OS. Windows installation is not supported.
 
+CCL works on Linux or Mac OS. Windows installation is not supported.
```
Hmm, not sure what's going on here but there seems to be a line duplicated.

mishakb replied:
fixed

c-d-leonard commented on src/ccl_background.c:
```diff
@@ -131,29 +131,51 @@ double ccl_rho_x(ccl_cosmology * cosmo, double a, ccl_species_x_label label, int
+	    s2_k = cosmo->params.lambda_mg*(hnorm*cosmo->params.H0)/k/(ccl_constants.CLIGHT/1000);
+        s1_k = (1.0+cosmo->params.c2_mg*s2_k*s2_k)/(1.0+s2_k*s2_k);
+
+//		printf("c1=%f,c2=%f,lambda=%f \n",cosmo->params.c1_mg,cosmo->params.c2_mg,cosmo->params.lambda_mg);
```
Should clean up these print statements at some point.

mishakb replied:
Done.

c-d-leonard commented on src/ccl_background.c:
```diff
@@ -196,16 +218,19 @@ static int growth_ode_system(double a,const double y[],double dydt[],void *param
 INPUT: scale factor
 TASK: Define the ODE system to be solved in order to compute the growth (of the density)
 * in the case in which we use the mu / Sigma quasistatic parameterisation of modified gravity
+MI: tag to get started 
```
Maybe this is a comment left for yourself?

mishakb replied:
removed.

c-d-leonard commented on src/ccl_halomod.c:
```diff
@@ -0,0 +1,360 @@
+#include <stdio.h>
```
looks like this file has perhaps been deleted on master since you last merged master into the branch? should be resolved by merging master in.

mishakb replied:
will do

mishakb replied:
now removed

c-d-leonard commented on src/ccl_musigma.c:
```diff
@@ -139,7 +139,7 @@ void ccl_cosmology_spline_linpower_musigma(ccl_cosmology* cosmo, ccl_f2d_t *psp,
           norm_pk, cosmo->params.n_s,
           cosmo->params.bcm_log10Mc, cosmo->params.bcm_etab,
-          cosmo->params.bcm_ks, 0., 0., cosmo->params.nz_mgrowth,
+          cosmo->params.bcm_ks, 0., 0., 0., 0., 0.,cosmo->params.nz_mgrowth,
```
do you want these at 0 by default? Somewhere else it looked like the default values should be 1.

mishakb replied:
Good catch. They should be 1., 1., 0. Done.

c-d-leonard commented on src/ccl_tracers.c:
```diff
@@ -348,14 +348,17 @@ void ccl_get_lensing_mag_kernel(ccl_cosmology *cosmo,
           chi = chi_arr[ichi];
           a = ccl_scale_factor_of_chi(cosmo, chi, &local_status);
           z = 1./a-1;
+          // mgfac = 1.0;
```
some of these commented-out lines probably need deleting?

mishakb replied:
removed.

c-d-leonard commented on src/ccl_tracers.c:
```diff
@@ -385,11 +388,15 @@ void ccl_get_kappa_kernel(ccl_cosmology *cosmo, double chi_source,
     double chi = chi_arr[ichi];
     double a = ccl_scale_factor_of_chi(cosmo, chi, status);
-    double mgfac = 1;
+ //   double mgfac = 1;
```
same thing here, don't forget to delete commented out lines.

mishakb replied:
removed too!

c-d-leonard commented:
@mishakb This is looking really great! I left a bunch of minor inline comments for you.

You may be aware of this but you do have some failing tests, so make sure to have a look at those.

I also think it's going to be quite important for this PR to update the note with the particular form of scale dependence you have implemented here.

Let me know when you're ready for another pass-through!

c-d-leonard commented:
Oh also you seem to be a bit behind master, probably merging that in sooner rather than later would be a good call.

mishakb commented:
Thanks very much @c-d-leonard for your useful comments. I will be working on them.

mishakb commented:
Thanks again @c-d-leonard for all the comments! I addressed most of them except a couple for the *.rst files that I will clean in the next few days as I will merge master into my branch so will fix any conflict at once. I also left a small question for you further above. Finally, I made progress on running the benchmarks, run successfully half of them and should finish them in the next few days!

c-d-leonard commented:
Glad this is progressing! I answered you questions above. Feel free to ping me again after you finish the thing mentioned above for another reivew.

damonge commented on pyccl/background.py:
```diff
@@ -271,35 +272,65 @@ def rho_x(cosmo, a, species, is_comoving=False):
-def Sig_MG(cosmo, a):
-    """Redshift-dependent modification to Poisson equation for massless
-    particles under modified gravity.
+def mu_MG(cosmo, a, k=None):
```
Since this is a modification to the Poisson equation it should probably not sit under `background`, but somewhere else.

c-d-leonard replied:
Yeah, that was my fault, I put those there in the first place. We should move them - ideas about where?

damonge replied:
If a lot of this is gonna be hidden under the hood in the API, I would suggest moving them to `tracers.py` and hiding them by changing the name into e.g. `_mu_MG`.

mishakb replied:
Fixed.  These were moved from backgound.py to musigma.py.

damonge commented on include/ccl_core.h:
```diff
@@ -333,7 +336,8 @@ ccl_parameters ccl_parameters_create(double Omega_c, double Omega_b, double Omeg
 				     double n_s, double bcm_log10Mc, double bcm_etab, double bcm_ks,
-				     double mu_0, double sigma_0, int nz_mgrowth, double *zarr_mgrowth,
+				     double mu_0, double sigma_0, double c1_mg, double c2_mg, double lambda_mg,
+                                     int nz_mgrowth, double *zarr_mgrowth,
```
indentation

mishakb replied:
fixed

mishakb replied:
Fixed. There was tabs instead of spaces in some of those lines and I suspect that that what was causing the problem but was not visible on the editor vi.

damonge commented on pyccl/background.py:
```diff
@@ -271,35 +272,65 @@ def rho_x(cosmo, a, species, is_comoving=False):
+                                  lib.mu_MG_vec,
+                                  cosmo, a, k)
+	
+def Sig_MG(cosmo, a, k=None):
```
Same comment about this being a `background` quantity

damonge replied:
As I said above, probably the best thing to do is move this to `tracers.py` and call it `_Sig_MG`.

mishakb replied:
Also moved to musigma.py

damonge commented on pyccl/tracers.py:
```diff
@@ -451,14 +540,15 @@ def __init__(self, cosmo, has_rsd, dndz, bias, mag_bias=None):
             t_a = (a_s, -growth_rate(cosmo, a_s))
             self.add_tracer(cosmo, kernel=kernel_d,
-                            transfer_a=t_a, der_bessel=2)
+                           transfer_a=t_a, der_bessel=2)         
```
flake will complain about this change in indentation

mishakb replied:
Done

damonge reviewed (changes requested):
OK, I've started a review but there are a few general issues we should discuss before going into the specifics.

First, if I understand correctly, the main point of this PR is making sure that we treat any quantities that depend on the matter overdensity and the Weyl potential consistently. Both power spectra would be the same in LCDM, but not in MG. Right now you are doing this by applying scale-dependent prefactors to the tracers. However, given the new capabilities of CCL to hold an arbitrary number of power spectra, I think it would be cleaner to apply these corrections at the power spectrum level, and store power spectra for matter x matter, matter x weyl and weyl x weyl, which then need to be combined with tracers (which know nothing about mu/sigma) to yield C_ells.

The other main comment is: before anything else, I think you should merge master into this branch, since the existing conflicts probably affect a lot of the new code in this PR.

mishakb commented:
OK. Thanks @damonge   for your comments. To start, I have merged master into the branch and resolved the conflicts.  As for your second point, I suggest we proceed in a sequential way which I think is fair to all :  I finish addressing your comments below which seem all reasonable and can be addressed pretty quickly (I will finish them tomorrow  (Thursday), then we merge this PR. We can then move on to the next step where someone else can creates and works on a new PR, implement and test MG as you describe. If all works and all are  happy with that then we can deprecate what is there, if that is the way we want to go.  I think this is fair to all. I will continue working on the comments.

damonge commented on pyccl/tracers.py:
```diff
@@ -533,15 +638,21 @@ class CMBLensingTracer(Tracer):
             is enough.
     """
-    def __init__(self, cosmo, z_source, n_samples=100):
+    def __init__(self, cosmo, z_source, n_samples=100, k=None):
```
Same comment as for `WeakLensingTracer`: explicit k dependence should be avoided

mishakb replied:
Done.

damonge commented on pyccl/tracers.py:
```diff
@@ -585,3 +696,157 @@ def _check_returned_tracer(return_val):
+    return mg_transfer
+
+
+    def MG_add_tracer(self, cosmo, kernel, mg_transfer, 
```
There seem to be two instances of this function. I suspect this second one should disappear

mishakb replied:
Done.

damonge commented on pyccl/tracers.py:
```diff
@@ -585,3 +696,157 @@ def _check_returned_tracer(return_val):
+
+
+
+def get_MG_transfer_function(cosmo, z, k=None):
```
I would also suggest making this a method of the `Tracer` class.

mishakb replied:
Done

damonge reviewed (changes requested):
OK, I've made a full pass through the code. There are a few things I haven't done because I think it'll be easier once these comments have been addressed (I've focused on the overall structure so far):
- I haven't looked at the tests
- I haven't looked at the inside main new functions (`get_MG_transfer_function` and `MG_add_tracer`).

It'd be very helpful to add documentation for the new parameters so we have a clear idea of what the new functionality is.

damonge commented:
(forgot to say: I promise I'll be quick with my next review. Sorry for the delay on this one)

mishakb commented on pyccl/core.py:
```diff
@@ -519,17 +525,17 @@ def _build_parameters(
+                   Omega_c, Omega_b, Omega_k, Neff,
+                   w0, wa, h, norm_pk,
+                   n_s, bcm_log10Mc, bcm_etab, bcm_ks,
+                   mu_0, sigma_0, c1_mg, c2_mg, lambda_mg, mnu_final_list, status)
```
thanks for adding them.

mishakb commented on pyccl/core.py:
```diff
@@ -519,17 +525,17 @@ def _build_parameters(
+                   Omega_c, Omega_b, Omega_k, Neff,
+                   w0, wa, h, norm_pk,
+                   n_s, bcm_log10Mc, bcm_etab, bcm_ks,
+                   mu_0, sigma_0, c1_mg, c2_mg, lambda_mg, z_mg, df_mg, mnu_final_list, status)
```
thanks for adding them

mishakb commented on readthedocs/api/pyccl.halos.rst:
```diff
@@ -7,12 +7,18 @@ Submodules
    pyccl.halos.massdef
    pyccl.halos.profiles
    pyccl.halos.profiles_2pt
+>>>>>>> master
```
now removed

mishakb commented on src/ccl_core.c:
```diff
@@ -471,6 +475,33 @@ ccl_parameters ccl_parameters_create(double Omega_c, double Omega_b, double Omeg
+
+}
+
+
```
now removed.

mishakb commented on src/ccl_musigma.c:
```diff
@@ -126,8 +126,15 @@ void ccl_rescale_musigma_s8(ccl_cosmology* cosmo, ccl_f2d_t *psp,
+          norm_pk, cosmo->params.n_s,
+          cosmo->params.bcm_log10Mc, cosmo->params.bcm_etab,
+          cosmo->params.bcm_ks, 0., 0., 1., 1., 0.,cosmo->params.nz_mgrowth,
+          cosmo->params.z_mgrowth, cosmo->params.df_mgrowth, status);
```
file now removed. Also, fixed the conflict for those lines -- now same as master.

mishakb commented on src/ccl_tracers.c:
```diff
@@ -310,14 +310,9 @@ void ccl_get_lensing_mag_kernel(ccl_cosmology *cosmo,
           ipar->chi_end = chi;
-
-          wL_arr[ichi] = lensing_kernel_integrate(cosmo, ipar, w)*(1+z)*lens_prefac*mgfac;
+          wL_arr[ichi] = lensing_kernel_integrate(cosmo, ipar, w)*(1+z)*lens_prefac;
```
yes.

mishakb commented on src/ccl_tracers.c:
```diff
@@ -347,11 +342,7 @@ void ccl_get_kappa_kernel(ccl_cosmology *cosmo, double chi_source,
-    if (fabs(cosmo->params.sigma_0))
-      mgfac += ccl_Sig_MG(cosmo, a, status);
-    wchi[ichi] = lens_prefac*(ccl_sinn(cosmo,chi_source-chi,status))*chi*mgfac/a;
+    wchi[ichi] = lens_prefac*(ccl_sinn(cosmo,chi_source-chi,status))*chi/a;
```
affirmative

mishakb commented on pyccl/tracers.py:
```diff
@@ -263,6 +262,95 @@ def get_bessel_derivative(self):
             return []
 
         return np.array([t.der_bessel for t in self._trc])
+#MI tagged 	
```
done

mishakb commented on pyccl/tracers.py:
```diff
@@ -263,6 +262,95 @@ def get_bessel_derivative(self):
 
         return np.array([t.der_bessel for t in self._trc])
+#MI tagged 	
+    def MG_add_tracer(self, cosmo, kernel, mg_transfer, 
```
done.

mishakb commented on pyccl/ccl_background.i:
```diff
@@ -6,6 +6,7 @@
 
 // Enable vectorised arguments for arrays
 %apply (double* IN_ARRAY1, int DIM1) {(double* a, int na)};
+%apply (double* IN_ARRAY1, int DIM1) {(double* k, int nk)};
```
moved to musigma.i

mishakb replied:
Oops. Removed now.

mishakb commented on pyccl/ccl_background.i:
```diff
@@ -78,26 +79,49 @@ void rho_x_vec(ccl_cosmology * cosmo, int label, int is_comoving,
-    for(int i=0; i < na; i++){
-      output[i] = ccl_Sig_MG(cosmo, a[i], status);
+%inline %{
+void mu_MG_vec(ccl_cosmology * cosmo, double* a, int na, double* k, int nk,
```
moved to musigma.i

mishakb commented on pyccl/ccl_background.i:
```diff
@@ -78,26 +79,49 @@ void rho_x_vec(ccl_cosmology * cosmo, int label, int is_comoving,
     }
 }
 
+void Sig_MG_vec(ccl_cosmology * cosmo, double* a, int na, double* k, int nk,
```
moved to musigma.i

mishakb commented on pyccl/core.py:
```diff
@@ -156,7 +156,8 @@ def __init__(
             bcm_log10Mc=np.log10(1.2e14), bcm_etab=0.5,
-            bcm_ks=55., mu_0=0., sigma_0=0., z_mg=None, df_mg=None,
+            bcm_ks=55., mu_0=0., sigma_0=0., 
+            c1_mg=1., c2_mg=1., lambda_mg=0., z_mg=None, df_mg=None,
```
removed old file. Documented the parameters in core.py

mishakb commented on pyccl/tracers.py:
```diff
@@ -451,14 +540,15 @@ def __init__(self, cosmo, has_rsd, dndz, bias, mag_bias=None):
             kernel_m = (chi, -2 * w)
             self.add_tracer(cosmo, kernel=kernel_m,
-                            der_bessel=-1, der_angles=1)
+                           der_bessel=-1, der_angles=1)   
```
Done -- file passes flake8

mishakb commented on src/ccl_background.c:
```diff
@@ -131,29 +131,48 @@ double ccl_rho_x(ccl_cosmology * cosmo, double a, ccl_species_x_label label, int
+        s1_k = (1.0+cosmo->params.c2_mg*s2_k*s2_k)/(1.0+s2_k*s2_k);
+
+	}
+	return cosmo->params.sigma_0 * ccl_omega_x(cosmo, a, ccl_species_l_label, status)/cosmo->params.Omega_l*s1_k;
```
same. They have been moved to ccl_musigma.c

mishakb commented on pyccl/tracers.py:
```diff
@@ -1,11 +1,10 @@
-from .background import comoving_radial_distance, growth_rate, \
-    growth_factor, scale_factor_of_chi
+from .background import comoving_radial_distance, growth_rate, \ 
+    growth_factor, scale_factor_of_chi, Sig_MG
```
Agreed and done.

mishakb commented on pyccl/tracers.py:
```diff
@@ -1,11 +1,10 @@
 from .pyutils import _check_array_params, NoneArr
 import numpy as np
 
-
```
fixed.

mishakb commented on pyccl/tracers.py:
```diff
@@ -483,7 +573,7 @@ class WeakLensingTracer(Tracer):
     """
     def __init__(self, cosmo, dndz, has_shear=True, ia_bias=None,
-                 use_A_ia=True):
+                 use_A_ia=True, k=None):
```
Agreed. Done.

mishakb commented:
Thanks very much @damonge and @c-d-leonard for all the good comments and I have addressed all of them. Looking forward to merging and tagging !

damonge commented on pyccl/tracers.py:
```diff
@@ -264,6 +265,157 @@ def get_bessel_derivative(self):
+            stepsize = z[1]-z[0]
+            samplesize = int(z[0]/stepsize)
+            z_0_to_zmin = np.linspace(0.0, z[0] - stepsize, samplesize)
+            z = np.concatenate((z_0_to_zmin, z))
```
Same question. Why do you need to go to `z=0`? If you do, it would be easier to just require this from the user, rather than forcing a specific sampling rate (or creating an array with 100 elements in the previous case)

mishakb replied:
It needs to go to zero so the transfer can be defined all the way to z = 0. Tried to use 100 elements but got some problems but this is not computationally demanding so I left it this way.

damonge commented on pyccl/tracers.py:
```diff
@@ -264,6 +265,157 @@ def get_bessel_derivative(self):
+        elif (isinstance(k, float) or isinstance(k, int)):
+            mgfac_1d = 1
+            mgfac_1d += Sig_MG(cosmo, a, k)
+        # setting transfer function with an array of a and a value of k
```
Comment should be indented

mishakb replied:
Done.

damonge commented on pyccl/tracers.py:
```diff
@@ -264,6 +265,157 @@ def get_bessel_derivative(self):
+            # computing MG factor
+            mgfac_1d = 1
+            mgfac_1d += Sig_MG(cosmo, a, k)
+        # setting transfer function with an array of a and k=None
```
comment should be indented

mishakb replied:
Done

damonge commented on pyccl/tracers.py:
```diff
@@ -264,6 +265,157 @@ def get_bessel_derivative(self):
+                                   bias_transfer_a[1])
+                self.add_tracer(cosmo, kernel, transfer_a=mg_transfer_new,
+                                der_bessel=der_bessel, der_angles=der_angles)
+            else:
```
I think these lines are equivalent to the previous case, so both cases could be merged into one.

mishakb replied:
condensed

mishakb replied:
condensed

damonge commented on pyccl/tracers.py:
```diff
@@ -264,6 +265,157 @@ def get_bessel_derivative(self):
+                self.add_tracer(cosmo, kernel, transfer_a=mg_transfer_new,
+                                der_bessel=der_bessel, der_angles=der_angles,
+                                transfer_k=bias_transfer_k)
+            else:
```
Can this case be merged with the previous one? I think all the lines are the same.

mishakb replied:
condensed

damonge commented on pyccl/tracers.py:
```diff
@@ -264,6 +265,157 @@ def get_bessel_derivative(self):
+                self.add_tracer(cosmo, kernel,
+                                transfer_a=mg_transfer, der_bessel=der_bessel,
+                                der_angles=der_angles)
+            else:
```
Can this case be merged with the previous one? I think all the lines are the same.

mishakb replied:
condensed

damonge commented on pyccl/tracers.py:
```diff
@@ -264,6 +265,157 @@ def get_bessel_derivative(self):
+                self.add_tracer(cosmo, kernel, transfer_a=mg_transfer,
+                                der_bessel=der_bessel, der_angles=der_angles,
+                                transfer_k=bias_transfer_k)
+            else:
```
As before, I think this can be merged with the previous one

mishakb replied:
condensed

damonge commented on pyccl/tracers.py:
```diff
@@ -518,8 +678,17 @@ def __init__(self, cosmo, dndz, has_shear=True, ia_bias=None,
+            else:
+                # MG case
+                k = np.logspace(-5, -0.7)
+                mg_transfer = self._get_MG_transfer_function(cosmo, z_a, k)
```
Since intrinsic alignments trace the matter potential rather than the "light" potential, should we actually be modifying this tracer function?

damonge replied:
And the converse of this: shouldn't the magnification part of number counts be modified (since it's related to lensing)?

mishakb replied:
Agreed. All fixed.

damonge reviewed (changes requested):
Thanks @mishakb 
More comments below. I've looked at the structure of the new code.

My main confusion is what the role of `mu_MG` is, since it doesn't seem to be used anywhere (I may be confused), and also which contributions of the tracers should take MG modifications and which ones shouldn't.

As soon as this is addressed I'll look at the tests.

c-d-leonard commented on pyccl/core.py:
```diff
@@ -518,18 +530,41 @@ def _build_parameters(
+                                                                    lambda_mg,
+                                                                    z_mg,
+                                                                    df_mg,
+                                                                    mnu_final_list,
```
flake8 seems to be failing on github actions because of this line being too long. I'm now sure how to solve this problem really since it's already a bit horrible other than making that variable name shorter. @damonge ?

damonge replied:
The previous version of this line had the arguments wrapped up nicely, so I'd just copy that (check the diff)

mishakb replied:
gave me trouble earlier but this time flake8 was nicer, it let me do what I wanted, so this is fixed.

mishakb commented:
Thanks @damonge for your responses and the new comments -- I am working on them.

mishakb commented on modified_files:
```diff
@@ -0,0 +1,8 @@
+git add   include/ccl_background.h
```
removed

mishakb commented on pyccl/__init__.py:
```diff
@@ -60,6 +60,9 @@
 from .neutrinos import Omeganuh2, nu_masses
 
+# MG functions
+from .musigma import mu_MG, Sig_MG
```
These have been now moved to tracers.py. and made private with a _

mishakb commented on pyccl/core.py:
```diff
@@ -132,6 +132,12 @@ class Cosmology(object):
+        c2_mg (:obj:`float`, optional): One of the parameters of the scale
+            dependent mu-Sigma modified gravity model. Defaults to 1.0
+        lambda_mg (:obj:`float`, optional): One of the parameters of the scale
+            dependent mu-Sigma modified gravity model. Defaults to 0.0
```
k-array removed. I now made the documentation more description and different.

mishakb commented on pyccl/musigma.py:
```diff
@@ -0,0 +1,71 @@
+"""
```
file removed and function moved to  tracers.py

mishakb commented on pyccl/tracers.py:
```diff
@@ -2,6 +2,7 @@
 from .core import check
 from .background import comoving_radial_distance, growth_rate, \
     growth_factor, scale_factor_of_chi
+from .musigma import Sig_MG
```
Removed. mu_MG was initially thought to serve for the growth at the level of Python but was not needed at the end.  So it have now been removed.

mishakb commented on src/ccl_background.c:
```diff
@@ -200,13 +174,15 @@ TASK: Define the ODE system to be solved in order to compute the growth (of the
-  double mu = ccl_mu_MG(cosmo, a, &status);
-  dydt[1]=1.5*hnorm*a*om*y[0]*(1. + mu);
+  double mu = ccl_mu_MG(cosmo, a, k, &status);
+  dydt[1]=1.5*hnorm*a*om*y[0]*(1.);// + mu);
```
Not sure what happened. All restored now.

mishakb commented on pyccl/ccl_musigma.i:
```diff
@@ -0,0 +1,55 @@
+                _status = 0;
+                output[i+index] = ccl_mu_MG(cosmo, a[i], k[j], &_status);
+                *status |= _status;
+// printf("index-i=%d, index-j=%d, index=%d, Sigma=%f, a=%f, k=%f \n", i, j, index, output[i+index], a[i], k[j]);
```
As I commented above. It was tabs that were messing things up. I believe I got it fixed.   
Debugging line removed.

mishakb commented on pyccl/ccl_musigma.i:
```diff
@@ -0,0 +1,55 @@
+                _status = 0;
+                output[i+index] = ccl_Sig_MG(cosmo, a[i], k[j], &_status);
+                *status |= _status;
+//printf("index-i=%d, index-j=%d, index=%d, Sigma=%f, a=%f, k=%f \n", i, j, index, output[i+index], a[i], k[j]);
```
Removed.

mishakb commented on src/ccl_musigma.c:
```diff
@@ -11,6 +11,51 @@
+    }
+    else {
+	    s2_k = cosmo->params.lambda_mg*(hnorm*cosmo->params.H0)/k/(ccl_constants.CLIGHT/1000);
+        s1_k = (1.0+cosmo->params.c2_mg*s2_k*s2_k)/(1.0+s2_k*s2_k);
```
Not necessarily ISiTGR has it but hi_class probably not. I am not sure  I understand all your questions but this is probably most used redshift and scale dependent parametrization for mu/Sigma although this is an area of active development . When you will want to discuss this future API I would be happy to discuss it with you.

mishakb commented on pyccl/tracers.py:
```diff
@@ -495,10 +648,17 @@ def __init__(self, cosmo, dndz, has_shear=True, ia_bias=None,
+                                der_bessel=-1, der_angles=2)
+            else:
+                # MG case
+                k = np.logspace(-5, -0.7)
```
see my response below.

damonge replied:
You could even embed this code in `_get_MG_transfer_function` and have it return both `mg_transfer` and `k`. That way you don't have to do this every time.

I would even suggest calling `_get_MG_transfer_function` from `_MG_add_tracer`, since that would also simplify the code a lot.

mishakb commented on pyccl/tracers.py:
```diff
@@ -518,8 +678,17 @@ def __init__(self, cosmo, dndz, has_shear=True, ia_bias=None,
+                                der_bessel=-1, der_angles=2)
+            else:
+                # MG case
+                k = np.logspace(-5, -0.7)
```
Can not use the k sampling based on ells because these are specified and used at the level of the Cls that comes after the tracers are defined. However, the above is not an issue because this range is chosen to be good enough for   the interpolation of the MG parameters and also does not go beyond the linear regime.

damonge replied:
See other comments.

mishakb commented on pyccl/tracers.py:
```diff
@@ -538,9 +707,14 @@ def __init__(self, cosmo, z_source, n_samples=100):
+        if (cosmo['sigma_0'] == 0):
+            self.add_tracer(cosmo, kernel=kernel, der_bessel=-1, der_angles=1)
+        else:
+            k = np.logspace(-5, -0.7)
```
see response above

damonge replied:
see my comment above about the default k values

mishakb commented on pyccl/tracers.py:
```diff
@@ -264,6 +265,157 @@ def get_bessel_derivative(self):
+                                transfer_k=bias_transfer_k)
+
+    def _get_MG_transfer_function(self, cosmo, z, k=None):
+        """ This function allows to obtain the function Sigma(z,k) (1 or 2D
```
Removed. See response further above.

mishakb commented on pyccl/tracers.py:
```diff
@@ -264,6 +265,157 @@ def get_bessel_derivative(self):
+        """
+        if isinstance(z, float):
+            a_single = 1/(1+z)
+            a = np.linspace(a_single, 1, 100)
```
this is to sample scale factor from a very small (at CMB for example) all the way to 1 here today for the transfer function.  For  a < a_single it is GR. no MG

mishakb commented on pyccl/tracers.py:
```diff
@@ -264,6 +265,157 @@ def get_bessel_derivative(self):
+
+    def _get_MG_transfer_function(self, cosmo, z, k=None):
+        """ This function allows to obtain the function Sigma(z,k) (1 or 2D
+            arrays) for an array of redshifts (i.e. scale factors) coming from
```
done

mishakb commented:
@damonge I addressed all the new comments.  Unit tests for mu and Sigma with scale factor and wavenumber k were done previously so all is ready for you for a last pass (hopefully :) )!

mishakb commented:
Just a friendly ping @damonge. Thanks.

damonge commented on src/ccl_musigma.c:
```diff
@@ -144,6 +190,8 @@ void ccl_rescale_musigma_s8(ccl_cosmology* cosmo, ccl_f2d_t *psp,
         double renorm = D_mu/D_GR;
         rescale_factor[i] *= renorm*renorm;
+        //printf("D_GR=%f\n", D_GR);
+        //printf("D_mu=%f\n", D_mu);
```
Remove these comments

mishakb replied:
removed

mishakb replied:
Good suggestions and thanks for the default sampling code lines.  All done as suggested.

damonge commented on pyccl/boltzmann.py:
```diff
@@ -248,6 +248,9 @@ def get_isitgr_pk_lin(cosmo):
     cp.Sigma0 = cosmo['sigma_0']
+    cp.c1 = cosmo['c1_mg']
+    cp.c2 = cosmo['c2_mg']
+    cp.Lambda = cosmo['lambda_mg']
```
When adding these parameters to isitgr, will the power spectrum already contain these modifications? In which case, should we not correct for them in the tracers? I.e. are these applied to the matter power spectrum used by CCL?

mishakb replied:
ISiTGR matter power spectrum already has the effect of mu but we are not using the Weyl power spectrum so the effects of Sigma is needed in tracers.py.

damonge commented on examples/MG_redshift_only_dependence_examples.ipynb:
```diff
@@ -0,0 +1,558 @@
+{
```
This file should disappear from the repo before we merge.

mishakb replied:
removed

damonge commented on examples/MG_redshift_and_scale_dependences_examples.ipynb:
```diff
@@ -0,0 +1,491 @@
+{
```
This file should disappear from the repo before we merge.

mishakb replied:
removed

damonge commented on benchmarks/test_correlation_MG3_SI.py:
```diff
@@ -0,0 +1,157 @@
+                          transfer_function='boltzmann_isitgr',
+                          matter_power_spectrum='linear')
+    cosmo.cosmo.gsl_params.INTEGRATION_LIMBER_EPSREL = 2.5E-5
+    cosmo.cosmo.gsl_params.INTEGRATION_EPSREL = 2.5E-5
```
Can you explain what are the differences between this test and the one in https://github.com/LSSTDESC/CCL/blob/ScaleDependent_muSigma/benchmarks/test_correlation_MG2.py?

mishakb replied:
none. I was just making sure that what I do will still keep the previous MG implementation working. But now I know it works so I just removed MG3 one so no duplicate.

damonge commented on benchmarks/test_correlation_MG3_SD.py:
```diff
@@ -0,0 +1,158 @@
+
+    # Read benchmarks
+    bms = {}
+    bms['dd_11'] = np.loadtxt(dirdat+'/wtheta_linear_predictionSD.dat')[0:15]
```
I don't think the scale-dependent benchmark files have been uploaded to the repo.

Also, how were the benchmarks created?

mishakb replied:
Not sure what happened. I thought I did but apparently the data files stayed behind (due to the ignore). I just uploaded all of them with add -f so it worked.  The benchmarks where created using ISiTGR.

mishakb replied:
Ok, thanks @damonge for the third batch of comments and I have addressed all them as well as the new suggestions. 
I also addressed @c-d-leonard comments and the last ones about flake8. 
Thanks to both of you for all the comments and suggestions. I hope we are ready to move forward.

damonge reviewed (changes requested):
OK @mishakb . Almost converged!

mishakb commented:
Ok @damonge, I am on them.

mishakb commented on pyccl/__init__.py:
```diff
@@ -39,7 +39,7 @@
 # Cl's and tracers
 from .tracers import Tracer, NumberCountsTracer, WeakLensingTracer, CMBLensingTracer, \
-    tSZTracer, get_density_kernel, get_kappa_kernel, get_lensing_kernel
+    tSZTracer, get_density_kernel, get_kappa_kernel, get_lensing_kernel, _Sig_MG
```
done

mishakb commented on pyccl/core.py:
```diff
@@ -132,6 +132,12 @@ class Cosmology(object):
+        c2_mg (:obj:`float`, optional): MG parameter that enters in the scale
+            dependence of Sigma affecting its large scale behavior. Default 1.
+        lambda_mg (:obj:`float`, optional): MG parameter that sets the start
+            of dependance on c1 and c2 MG parameters. Defaults to 0.0
```
Done. Equations and paper referenced.

damonge replied:
Great, thanks a lot! Can you also specify what is assumed about f_1 and f_2 in those equations?

mishakb replied:
Very good question and I agree that completes the definition, thanks! I added a clear statement for f1 and f2.

mishakb commented on pyccl/core.py:
```diff
@@ -160,7 +166,8 @@ def __init__(
             bcm_log10Mc=np.log10(1.2e14), bcm_etab=0.5,
-            bcm_ks=55., mu_0=0., sigma_0=0., z_mg=None, df_mg=None,
+            bcm_ks=55., mu_0=0., sigma_0=0.,
+            c1_mg=1., c2_mg=1., lambda_mg=0., z_mg=None, df_mg=None,
```
affirmative

mishakb commented on src/ccl_musigma.c:
```diff
@@ -11,6 +11,51 @@
+	    s2_k = (cosmo->params.lambda_mg*(hnorm*cosmo->params.H0)/k/(ccl_constants.CLIGHT/1000));
+	    s1_k = (1.0+cosmo->params.c1_mg*s2_k*s2_k)/(1.0+s2_k*s2_k);
+	}
+	return cosmo->params.mu_0 * ccl_omega_x(cosmo, a, ccl_species_l_label, status)/cosmo->params.Omega_l*s1_k;
```
You are right but I would like to suggest to keep it for now just keeps the code balanced and in case we need it for something else before we put the new API. I hope you do not have a strong feeling about this one and will let me keep this.

damonge replied:
No, that's fine!

mishakb commented on pyccl/tracers.py:
```diff
@@ -458,8 +607,17 @@ def __init__(self, cosmo, has_rsd, dndz, bias, mag_bias=None):
+                                der_bessel=-1, der_angles=1)
+            else:
+                # MG case
+                k = np.logspace(-5, -0.7)
```
Ok, added the if statement, commented to explain what is done, and the k sampling is replaced as suggested below.

damonge commented on pyccl/tracers.py:
```diff
@@ -264,6 +296,133 @@ def get_bessel_derivative(self):
+
+        # case with no astro biases
+        if ((bias_transfer_a is None) and (bias_transfer_k is None)):
+            if (isinstance(k, (list, np.ndarray))):
```
OK, now we know that `k` is always an array, right? In which case, do you need these `if` statements?
It's good to avoid unnecessary `if`s because otherwise coveralls will complain that you're not testing all possible if-branches of the code.

mishakb replied:
For the further above comment: Very good question and I agree that completes the definition, thanks! I added a clear statement for f1 and f2.

I also agree for the unnecessary ifs -- all removed!

damonge commented on pyccl/tracers.py:
```diff
@@ -264,6 +296,133 @@ def get_bessel_derivative(self):
+        # add.tracer
+        mgfac_2d = mgfac_1d.reshape(len(a), -1, order='F')
+        # converting to ln(k) array
+        lk = np.log(k)
```
ah, I didn't realise you were going to need `log(k)`. In that case, that's what `get_pk_spline_lk` returns directly, so you can do this more efficiently by changing the lines above and removing this one (see my previous comment)

mishakb replied:
Good point -- done!

damonge commented on pyccl/tracers.py:
```diff
@@ -264,6 +296,133 @@ def get_bessel_derivative(self):
+        status = 0
+        k, status = lib.get_pk_spline_lk(cosmo.cosmo, nk, status)
+        check(status)
+        k = np.exp(k)
```
If you change the previous lines to:
```
        lk, status = lib.get_pk_spline_lk(cosmo.cosmo, nk, status)
        check(status)
        k = np.exp(lk)
```
you'll have log(k) already computed (so you save on calculating one extra spurious logarithm). See my next comment.

mishakb replied:
Agreed. Done!

damonge reviewed (changes requested):
Thanks @mishakb . Almost there, one last small batch of tiny comments!

mishakb commented:
Ok, @damonge, thanks a lot for the last comments and they have been all addressed.  Could you approve?
I also had addressed all the comments from @c-d-leonard.  Thanks to both of you for the good reviews. I would like to merge asap and rest :)

damonge commented:
@mishakb some of the tests aren't passing (I think because this PR changes the way params are passed around). It's probably worth running the tests locally to speed up the debugging.

mishakb commented:
@damonge, yes I have been fixing those. Did not realize that those can be run using pytest as the others so I am now running them locally. That will go faster. Please keep checking from time to time -- maybe we can get this finished today.

mishakb commented:
@damonge, all tests have passed!  could you approve :)

damonge reviewed (approved):
OK, ready to go!
